cmd/enterprise-portal/internal/database/subscriptions.go

```diff
@@ -1,10 +1,164 @@
 	// "acme.sourcegraphcloud.com".
-	InstanceDomain string `gorm:"index"`
+	InstanceDomain string `gorm:"unique"`
+}
+
+func (s *Subscription) TableName() string {
+	return "enterprise_portal_subscriptions"
+}
+
+// SubscriptionsStore is the storage layer for product subscriptions.
+type SubscriptionsStore struct {
+	db *pgxpool.Pool
+}
+
+func newSubscriptionsStore(db *pgxpool.Pool) *SubscriptionsStore {
+	return &SubscriptionsStore{
+		db: db,
+	}
+}
+
+// ListSubscriptionsOptions is the set of options to filter subscriptions.
+// Non-empty fields are treated as AND-concatenated.
+type ListSubscriptionsOptions struct {
+	// IDs is a list of subscription IDs to filter by.
+	IDs []string
+	// InstanceDomains is a list of instance domains to filter by.
+	InstanceDomains []string
+	// OnlyArchived indicates whether to only list archived subscriptions.
+	OnlyArchived bool
+	// PageSize is the maximum number of subscriptions to return.
+	PageSize int
+}
+
+func (opts ListSubscriptionsOptions) toQueries() (whereConds, limit *sqlf.Query) {
+	conds := []*sqlf.Query{
+		sqlf.Sprintf("TRUE"),
+	}
+	if len(opts.IDs) > 0 {
+		subconds := dbutil.ArgsToQueries(opts.IDs)
+		conds = append(conds, sqlf.Sprintf("id IN (%s)", sqlf.Join(subconds, ", ")))
+	}
+	if len(opts.InstanceDomains) > 0 {
+		subconds := dbutil.ArgsToQueries(opts.InstanceDomains)
+		conds = append(conds, sqlf.Sprintf("id IN (%s)", sqlf.Join(subconds, ", ")))
+	}
+	if opts.OnlyArchived {
+		conds = append(conds, sqlf.Sprintf("archived = TRUE"))
+	}
+	whereConds = sqlf.Sprintf("(%s)", sqlf.Join(conds, " AND "))
+
+	limit = sqlf.Sprintf("")
+	if opts.PageSize > 0 {
+		limit = sqlf.Sprintf("LIMIT %d", opts.PageSize)
+	}
+	return whereConds, limit
+}
+
+// List returns a list of subscriptions based on the given options.
+func (s *SubscriptionsStore) List(ctx context.Context, opts ListSubscriptionsOptions) ([]*Subscription, error) {
+	whereConds, limit := opts.toQueries()
+	q := sqlf.Sprintf(`
+SELECT
+	id,
+	instance_domain
+FROM subscriptions
+WHERE %s
+%s`,
+		whereConds, limit,
+	)
+	rows, err := s.db.Query(ctx, q.Query(sqlf.PostgresBindVar), q.Args()...)
+	if err != nil {
+		return nil, errors.Wrap(err, "query rows")
+	}
+	defer rows.Close()
+
+	var subscriptions []*Subscription
+	for rows.Next() {
+		var subscription Subscription
+		if err = rows.Scan(&subscription.ID, &subscription.InstanceDomain); err != nil {
+			return nil, errors.Wrap(err, "scan row")
+		}
+		subscriptions = append(subscriptions, &subscription)
+	}
+	return subscriptions, rows.Err()
+}
+
+type UpsertSubscriptionOptions struct {
+	InstanceDomain string
+	// ForceUpdate indicates whether to force update all fields of the subscription
+	// record.
+	ForceUpdate bool
+}
+
+func (opts UpsertSubscriptionOptions) toQuery(id string) (_ *sqlf.Query, noop bool) {
+	if opts.ForceUpdate {
+		return sqlf.Sprintf(`
+INSERT INTO subscriptions (id, instance_domain)
+VALUES (%s, %s)
```

Comment:
With native pgx:

```suggestion
INSERT INTO subscriptions (id, instance_domain)
VALUES (@instance_domain, @id)
```

then, we return `pgx.NamedArgs`

Reply:
(also, is this a bug? the parameters are currently ordered `opts.InstanceDomain, id`)

Reply:
Yes it is a bug haha, fixed along with another comment.